allauth/socialaccount/providers/apple/client.py

```diff
@@ -0,0 +1,84 @@
+import requests
+from datetime import datetime, timedelta
+from urllib.parse import parse_qsl
+
+from django.utils.http import urlencode
+
+import jwt
+
+from allauth.socialaccount.providers.oauth2.client import (
+    OAuth2Client,
+    OAuth2Error,
+)
+
+
+class Scope(object):
+    EMAIL = 'email'
+    NAME = 'name'
+
+
+class AppleOAuth2Client(OAuth2Client):
+    """
+    Custom client because `Sign In With Apple`:
+        * requires `response_mode` field in redirect_url
+        * requires special `client_secret` as JWT
+    """
+
+    def generate_client_secret(self):
+        """Create a JWT signed with an apple provided private key"""
+        now = datetime.utcnow()
+        claims = {
+            'iss': self.key,
+            'aud': 'https://appleid.apple.com',
+            'sub': self.get_client_id(),
+            'iat': now,
+            'exp': now + timedelta(hours=1),
+        }
+        headers = {'kid': self.consumer_secret, 'alg': 'ES256'}
+        client_secret = jwt.encode(
+            payload=claims, key=self.cert, algorithm='ES256', headers=headers
+        ).decode('utf-8')
+        return client_secret
+
+    def get_client_id(self):
+        """ We support multiple client_ids, but use the first one for api calls """
+        return self.consumer_key.split(',')[0]
+
+    def get_access_token(self, code):
+        url = self.access_token_url
```

Comment:
I think we need to verify the token before getting the access token as it is mentioned in the [doc](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/verifying_a_user#overview). Correct me if I am wrong

Reply:
In that doc it says "Web apps must obtain the identity token from their server by validating the authorization code" so I think they're kind of saying that we're verifying the code by exchanging it for the access_token? That was my reading of it anyway. There are no other API endpoints that seem to fit the bill for verifying the code.

Reply:
@benhowes It is interesting. The docs are very much misleading with their terminology.
The code we are getting here is actually "authorization code" as mentioned in the [doc](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms) and it is not **Identity Token**


![image](https://user-images.githubusercontent.com/7114151/78423099-d553fa80-7685-11ea-8a4f-3cecdbb88b2e.png)


And the token we are getting after calling the `access_token_url` is besically `refresh token` as mentioned in the [doc](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/verifying_a_user#3385187). As per the doc "**After verifying the identity token on your server, call the Generate and validate tokens endpoint with the client_id, client_secret, and nonce information**", but we are trying to get the refresh token before validing the **`identity token`**

Reply:
@safwanrahman I just can say, that docs were very poor at December when I published this PR :)

Reply:
@0leg5ergeev Yeah, I can endorse that the docs were worse and I was almost gone while implementing sign in with apple with a short deadline back in december or january. It is good that it is better.